src/mscorlib/shared/System/String.cs

```diff
@@ -2,109 +2,391 @@
+
+            return CreateStringForSByteConstructor(pStart, length);
+        }
+
+        // Encoder for String..ctor(sbyte*) and String..ctor(sbyte*, int, int)
+        private static unsafe string CreateStringForSByteConstructor(byte *pb, int numBytes)
+        {
+            Debug.Assert(numBytes >= 0);
+            Debug.Assert(pb <= (pb + numBytes));
+
+            if (numBytes == 0)
+                return Empty;
+
+#if PLATFORM_UNIX
+            return Encoding.UTF8.GetString(pb, numBytes);
+#else
+            int numCharsRequired = Interop.Kernel32.MultiByteToWideChar(Interop.Kernel32.CP_ACP, Interop.Kernel32.MB_PRECOMPOSED, pb, numBytes, (char*)null, 0);
+            if (numCharsRequired == 0)
+                throw new ArgumentException(SR.Arg_InvalidANSIString);
+
+            string newString = FastAllocateString(numCharsRequired);
+            fixed (char *pFirstChar = &newString._firstChar)
+            {
+                numCharsRequired = Interop.Kernel32.MultiByteToWideChar(Interop.Kernel32.CP_ACP, Interop.Kernel32.MB_PRECOMPOSED, pb, numBytes, pFirstChar, numCharsRequired);
+            }
+            if (numCharsRequired == 0)
+                throw new ArgumentException(SR.Arg_InvalidANSIString);
+            return newString;
+#endif
+        }
+
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public extern unsafe String(sbyte* value, int startIndex, int length, Encoding enc);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private unsafe string Ctor(sbyte* value, int startIndex, int length, Encoding enc)
+        {
+            if (enc == null)
+                return new string(value, startIndex, length);
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            if (startIndex < 0)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
 
-                    // uppercase - notice that we need just one compare
-                    if ((uint)(c - 'a') <= (uint)('z' - 'a')) c -= 0x20;
+            byte* pStart = (byte*)(value + startIndex);
 
-                    outBuff[i] = (char)c;
+            // overflow check
+            if (pStart < value)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_PartialWCHAR);
+
+            return enc.GetString(new ReadOnlySpan<byte>(pStart, length));
+        }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        public extern String(char c, int count);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private string Ctor(char c, int count)
+        {
+            if (count > 0)
+            {
+                string result = FastAllocateString(count);
+
+                if (c == '\0')
```

Comment:
FastAllocateString explicitly zeros?  Or is that just behavior picked up from the allocator in general?  Since the common case for FastAllocateString is that we're going to overwrite the whole thing, I'm wondering if there's an opportunity to change it to not spend time zero'ing, and then we can clear only if we need to.

Reply:
> it is that just behavior picked up from the allocator in general?

Correct. Everything allocated on GC heap starts as zeroed out. The memory is zeroed out in chunks, so by the time we know that we want to allocate a (small) string, the memory for it has been typically zeroed already. We maybe able to skip the zeroing for large (large then the chunk) strings.

Reply:
The same optimization can be done for large arrays allocated via ArrayPool. It would be a lot more beneficial there, I think.

Reply:
Yup, that'd be welcome.